servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -598,21 +691,121 @@ public Completable onClose() {
         @SuppressWarnings("unchecked")
         private Completable doClose(final Function<? super C, Completable> closeFunction) {
             return Completable.defer(() -> {
-                final Object[] connections = connStateUpdater.getAndSet(this, CLOSED_CONN_STATE).connections;
+                final ConnState oldState = connStateUpdater.getAndSet(this, CLOSED_CONN_STATE);
+                cancelIfHealthCheck(oldState.state);
+                final Object[] connections = oldState.connections;
                 return connections.length == 0 ? completed() :
                         from(connections).flatMapCompletableDelayError(conn -> closeFunction.apply((C) conn));
             });
         }
 
+        private void cancelIfHealthCheck(Object o) {
+            if (HealthCheck.class.equals(o.getClass())) {
+                @SuppressWarnings("unchecked")
+                HealthCheck<Addr, C> healthCheck = (HealthCheck<Addr, C>) o;
+                LOGGER.debug("Health check for address {} cancelled.", healthCheck.host.address);
+                healthCheck.cancel();
+            }
+        }
+
         @Override
         public String toString() {
             final ConnState connState = this.connState;
             return "Host{" +
                     "address=" + address +
-                    ", state=" + connState.state +
+                    ", state=" + describeState(connState.state) +
                     ", #connections=" + connState.connections.length +
                     '}';
         }
+
+        private String describeState(Object state) {
+            if (ActiveState.class.equals(state.getClass())) {
+                return "active";
```

Comment:
Will be useful to implement `toString()` for `ActiveState` to see the current failure counter. Then you can simplify this method to:
```java
if (HealthCheck.class.equals(state.getClass())) {
    return "unhealthy";
} else {
    state.toString();
}
```